syntax = "proto3";
package fx.dex;

import "gogoproto/gogo.proto";
import "fx/dex/order.proto";

option go_package = "git.wokoworks.com/blockchain/fx-chain/x/dex/types";

service Msg {
    rpc CreateOrder(MsgCreateOrder) returns (MsgCreateOrderResponse);
    rpc CancelOrder(MsgCancelOrder) returns (MsgCancelOrderResponse);

    rpc AddMargin(MsgAddMargin) returns (MsgAddMarginResp);
    rpc ReduceMargin(MsgReduceMargin) returns (ReduceMarginResp);
    rpc ClosePosition(MsgClosePosition) returns (MsgClosePositionResp);
    rpc LiquidationPosition(MsgLiquidationPosition) returns (MsgLiquidationPositionResp);

    rpc CreatePair(MsgCreatePairRequest) returns (MsgCreatePairResponse);
}

message MsgCreateOrder {
    bytes owner = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    string pair_id = 2;
    Direction direction = 3;
    string price = 4 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
    string base_quantity = 5 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
    int64 ttl = 6;
    int64 leverage = 7;
}

message MsgCreateOrderResponse {
    string order_id = 1;
}

message MsgCancelOrder {
    bytes owner = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    string order_id = 2;
}

message MsgCancelOrderResponse {
}

message MsgAddMargin {
    bytes owner = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    string pair_id = 2;
    string position_id = 3 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Uint",
        (gogoproto.nullable) = false
    ];
    string margin = 4 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
}

message MsgAddMarginResp {
}

message MsgReduceMargin {
    bytes owner = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    string pair_id = 2;
    string position_id = 3 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Uint",
        (gogoproto.nullable) = false
    ];
    string margin = 4 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
}

message ReduceMarginResp {
}

message MsgClosePosition {
    bytes owner = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    string pair_id = 2;
    string position_id = 3 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Uint",
        (gogoproto.nullable) = false
    ];
    string price = 4 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
    string base_quantity = 5 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
    bool full_close = 6 ;
}

message MsgClosePositionResp {
}

message MsgLiquidationPosition {
    bytes liquidator = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
    string position_id = 2 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Uint",
        (gogoproto.nullable) = false
    ];
}

message MsgLiquidationPositionResp {
}


message MsgCreatePairRequest {
    string base_asset = 1;
    string quote_asset = 2;
    string validator = 3;
    string init_price = 4 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable) = false
    ];
}

message MsgCreatePairResponse {

}


